helm-v3/006-repositories.md

```diff
@@ -1,40 +1,153 @@
-> mechanisms.
-> 
-> [name=Adnan Abdulhussein] What would the default HTTP(S) implementation be?
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
+
+Afterwards, running a `helm repo add` pointing to any repository URL
+tied to a provider that has been successfully authenticated against will
+not require additional credentials.
+
+The exact semantics are TBD, but several auth methods will be supported,
+such as:
+
+* Basic Auth
+* Cert based Auth
+* Token/Bearer Auth (with token refresh)
+* Two-Factor Auth
+
+## The `helm push` command
```

Comment:
Will the helm push support upload both the chart package, as well as certain metadata? The metadata could be used to construct the chart level json file. I feel we should separate these two when doing a helm push so that the server implementation will not need to depend on the chart package to populate the metadata. This has advantage on both performance (server doesn't need to open the big package file) and flexibility (the metadata format can be changed/improved separately).

Reply:
I think that's a limitation on the helm chart package format, right? Might be a good proposal for https://github.com/helm/community/blob/master/helm-v3/001-charts.md.